psp-1.md

```diff
@@ -0,0 +1,459 @@
+			// The auction is available
+			ensure!(new_domain.available, "The auction for the domain is currently not available");
+			// The auction is not closed
+			let now = <timestamp::Module<T>>::now();
+			ensure!(new_domain.auction_closed > now, "The bid for the auction is already closed");
+			// The bid price is higher than the current highest bid
+			ensure!(new_domain.highest_bid < bid.clone(), "Bid higher");
+			
+
+			// Set new domain data
+			new_domain.bidder = sender.clone();
+			new_domain.highest_bid = bid.clone();
+			
+			// mutate domain with new_domain struct in the Domain state
+			<Resolver<T>>::mutate(domain_hash.clone(), |domain| *domain = new_domain.clone());
+			Self::deposit_event(RawEvent::NewBid(sender, domain_hash, bid));
+
+			Ok(())
+		}
+
+		pub fn finalize_auction(origin, domain_hash: T::Hash) -> Result {
+			let sender = ensure_signed(origin)?; 
+			// Ensure that
+			// Domain does already exist
+			ensure!(<Resolver<T>>::exists(domain_hash), "The domain is not registered yet");
+			// But wait, get domain data and time
+			let mut new_domain = Self::domain(domain_hash);
+			let now = <timestamp::Module<T>>::now();
+			// The auction is available
+			ensure!(new_domain.available, "The auction for the domain is currently not available");
+			// The auction is finalized or the source wants to finalize the auction(test)
+			// TEST: If you want to test auction functions without waiting for 1 hour, just add '|| sender == new_domain.source in ensure! macro
+			ensure!(now > new_domain.auction_closed, "The auction has not been finalized yet");
+
+			let _ = <balances::Module<T> as Currency<_>>::transfer(&new_domain.bidder, &new_domain.source, new_domain.highest_bid);
+
+			// Set new domain data to bidder as source, highest_bid as price, and reinitialize rest of them 
+			new_domain.source = new_domain.bidder.clone();
+			new_domain.price = new_domain.highest_bid;
+			new_domain.available = false;
+			let ttl = Self::to_milli(T::Moment::from(YEAR));
+			new_domain.ttl = ttl;
+			new_domain.registered_date = now;
+			new_domain.available = false;
+			new_domain.highest_bid = T::Balance::from(0);
+			new_domain.auction_closed = T::Moment::from(0);
+
+			// mutate domain with new_domain struct in the Domain state
+			<Resolver<T>>::mutate(domain_hash.clone(), |domain| *domain = new_domain.clone());
+			Self::deposit_event(RawEvent::AuctionFinalized(new_domain.bidder, domain_hash, new_domain.highest_bid));
+
+			Ok(())
+		}
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where AccountId = <T as system::Trait>::AccountId, <T as system::Trait>::Hash, <T as balances::Trait>::Balance, <T as timestamp::Trait>::Moment
+ {
+		DomainRegistered(AccountId, Balance, Moment, Moment),
+		NewAuction(AccountId, Hash, Moment, Moment), 
+		NewBid(AccountId, Hash, Balance),
+		AuctionFinalized(AccountId, Hash, Balance),
+		DomainResolved(Hash, AccountId),
+		DomainRenewal(Hash, AccountId, Moment),
```

Comment:
```suggestion
		DomainRenewed(Hash, AccountId, Moment),
```

Reply:
Thank you for your review. It was very interesting to review them.

1) I was thinking to make a hashmap in the Domain struct and use key as the chain index from [slips-44](https://github.com/satoshilabs/slips/blob/master/slip-0044.md), value as the byte array(Vec<u8>) of the corresponding blockchain address. I see some missing address schemes, so I might come up with the new index for polkadot extending the existing chain index. If you have other identifiers, please let me know.

2) I saw ENS has been working on [javascript client side library](https://github.com/ensdomains/address-encoder) to provide encoding/decoding service for the address in multi chains. I think I will make the library compatible with this standard. In fact, I think someone can help ENS's library adding polkadot address scheme as a [bounty](https://explorer.bounties.network/bounty/3584). The only difference would be the address string will be converted to a byte array(Vec<u8>) to store in Substrate. I think I should refer this too in the proposal.

3) Reverse lookup is possible, and I think I will use this data to check referral to encompass new users trying to use this service. Naming service requires a cryptocurrency to pay fee for interaction between blockchain. ENS requires ether, and the user must get one from crypto exchanges or his or her friend. I want to eliminate this barrier using native blockchain to let the general public use this service via referral. 

The process works like this:

1. User A makes a domain from the name service and registers its address to the Reverse lookup storage.

2. User B without any cryptocurrency balance wants to make a domain name, and User A sends the extrinsic that shows the intent that User B wants to register the domain.

3. The blockchain checks the Reverse lookup storage and checks User B exists in the Reverse lookup storage. If User B does not exist, the blockchain adds User B in the reverse lookup storage with a null value and mints 0.001 DOT(new domain registration fee) + extrinsic fees to User B. 

4. User B gets the fund to use domain name service and registers the new one for his or her account.